README.md
eos-speedwagon
==============

A classy startup spinner to stare at while applications load.



debian/changelog
eos-speedwagon (1.0.0-1) eos; urgency=low

  * Initial release.

 -- Jasper St. Pierre <jasper@endlessm.com>  Wed, 21 Jan 2015 18:23:38 -0800



autogen.sh
#!/bin/sh
# Run this to generate all the initial makefiles, etc.

srcdir=`dirname $0`
test -z "$srcdir" && srcdir=.

PKG_NAME="eos-speedwagon"
REQUIRED_AUTOCONF_VERSION=2.53
REQUIRED_AUTOMAKE_VERSION=${REQUIRED_AUTOMAKE_VERSION:-1.9}

(test -f $srcdir/speedwagon.gresource.xml) || {
    echo -n "**Error**: Directory "\`$srcdir\'" does not look like the"
    echo " top-level $PKG_NAME directory"
    exit 1
}

DIE=0

# This is a bit complicated here since we can't use gnome-config yet.
# It'll be easier after switching to pkg-config since we can then
# use pkg-config to find the gnome-autogen.sh script.

gnome_autogen=
gnome_datadir=

ifs_save="$IFS"; IFS=":"
for dir in $PATH ; do
  test -z "$dir" && dir=.
  if test -f $dir/gnome-autogen.sh ; then
    gnome_autogen="$dir/gnome-autogen.sh"
    gnome_datadir=`echo $dir | sed -e 's,/bin$,/share,'`
    break
  fi
done
IFS="$ifs_save"

if test -z "$gnome_autogen" ; then
  echo "You need to install the gnome-common module and make"
  echo "sure the gnome-autogen.sh script is in your \$PATH."
  exit 1
fi

GNOME_DATADIR="$gnome_datadir" USE_GNOME2_MACROS=1 . $gnome_autogen


src/js/main.js
// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-

const GLib = imports.gi.GLib;
const Gio = imports.gi.Gio;
const Gdk = imports.gi.Gdk;

// Needed to be imported so that get_window() returns a proper wrapper.
const GdkX11 = imports.gi.GdkX11;

const Gtk = imports.gi.Gtk;

const Format = imports.format;
const Lang = imports.lang;

const SpeedwagonIface = '<node> \
<interface name="com.endlessm.Speedwagon"> \
<method name="ShowSplash"> \
    <arg type="s" direction="in" name="desktopFile" /> \
</method> \
<method name="HideSplash"> \
    <arg type="s" direction="in" name="desktopFile" /> \
</method> \
<signal name="SplashClosed"> \
    <arg type="s" name="desktopFile" /> \
</signal> \
</interface> \
</node>';

const SPLASH_CIRCLE_PERIOD = 2;
const SPLASH_SCREEN_FADE_OUT = 1;

// This needs to match the time of the transitions in speedwagon.css
const SPLASH_SCREEN_COMPLETE_TIME = 1;

const SPLASH_BACKGROUND_DESKTOP_KEY = 'X-Endless-SplashBackground';
const DEFAULT_SPLASH_SCREEN_BACKGROUND = 'resource:///com/endlessm/Speedwagon/splash-background-default.jpg';

const Window = new Lang.Class({
    Name: 'EosSpeedwagonWindow',
    Extends: Gtk.ApplicationWindow,

    _init: function(app, appInfo) {
        this._appInfo = appInfo;

        this.parent({ application: app,
                      role: 'eos-speedwagon',
                      title: appInfo.get_name() });
        this.maximize();

        this.realize();
        let gdkWindow = this.get_window();
        let appID = appInfo.get_id().replace('.desktop', '');
        gdkWindow.set_utf8_property('_GTK_APPLICATION_ID', appID);

        this._buildUI();
    },

    _buildUI: function() {
        this._baseBox = new Gtk.Box({ visible: true });
        let context = this._baseBox.get_style_context();
        context.add_class('speedwagon-bg');

        this._spinner = new Gtk.Spinner({ visible: true,
                                          width_request: 220,
                                          height_request: 220,
                                          hexpand: true, vexpand: true,
                                          halign: Gtk.Align.CENTER,
                                          valign: Gtk.Align.CENTER });
        this._spinner.get_style_context().add_class('speedwagon-spinner');
        this._spinner.start();

        let spinnerBox = new Gtk.Box({ visible: true });
        spinnerBox.get_style_context().add_class('speedwagon-spinner-bg');

        let overlay = new Gtk.Overlay({ visible: true });
        overlay.add(spinnerBox);
        overlay.add_overlay(this._spinner);
        this._baseBox.add(overlay);

        let gicon = this._appInfo.get_icon();
        let image = new Gtk.Image({ visible: true,
                                    hexpand: true, vexpand: true,
                                    halign: Gtk.Align.CENTER,
                                    valign: Gtk.Align.CENTER,
                                    pixel_size: 64,
                                    gicon: gicon });
        spinnerBox.add(image);

        let bgPath;
        if (this._appInfo.has_key(SPLASH_BACKGROUND_DESKTOP_KEY)) {
            bgPath = this._appInfo.get_string(SPLASH_BACKGROUND_DESKTOP_KEY);
        } else {
            bgPath = DEFAULT_SPLASH_SCREEN_BACKGROUND;
        }

        let provider = new Gtk.CssProvider();
        provider.load_from_data(Format.vprintf('.speedwagon-bg { background-image: url("%s"); }', [bgPath]));
        context.add_provider(provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);

        this.add(this._baseBox);
    },

    _fadeOutTick: function(window, clock) {
        let time = clock.get_frame_time();
        let dt = (time - this._fadeOutStartTime) / GLib.USEC_PER_SEC;
        let opacity = 1.0 - (dt / SPLASH_SCREEN_FADE_OUT);
        if (opacity > 0)
            this.opacity = opacity;
        else
            this.destroy();
        return true;
    },
    _fadeOut: function() {
        this._fadeOutStartTime = this.get_frame_clock().get_frame_time();
        this.add_tick_callback(Lang.bind(this, this._fadeOutTick));
    },

    rampOut: function() {
        this._baseBox.get_style_context().add_class('glow');

        GLib.timeout_add_seconds(GLib.PRIORITY_DEFAULT, SPLASH_SCREEN_COMPLETE_TIME, Lang.bind(this, function() {
            this._fadeOut();
            return false;
        }));
    },
});

const EosSpeedwagon = new Lang.Class({
    Name: 'EosSpeedwagon',
    Extends: Gtk.Application,

    _init: function() {
        this.parent({ application_id: 'com.endlessm.Speedwagon' });

        this._dbusImpl = Gio.DBusExportedObject.wrapJSObject(SpeedwagonIface, this);
        this._dbusImpl.export(Gio.DBus.session, '/com/endlessm/Speedwagon');

        this._splashes = {};
    },

    vfunc_startup: function() {
        this.parent();

        let resource = Gio.Resource.load('speedwagon.gresource');
        resource._register();

        let provider = new Gtk.CssProvider();
        provider.load_from_file(Gio.File.new_for_uri('resource:///com/endlessm/Speedwagon/speedwagon.css'));
        Gtk.StyleContext.add_provider_for_screen(Gdk.Screen.get_default(), provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);

        this.hold();
    },

    vfunc_activate: function() {
    },

    _emitSplashClosed: function(appID) {
        let connection = this._dbusImpl.get_connection();
        let info = this._dbusImpl.get_info();
        connection.emit_signal(null, this._dbusImpl.get_object_path(), info.name, 'SplashClosed', GLib.Variant.new('(s)', [appID]));
    },

    ShowSplash: function(appID) {
        if (!this._splashes[appID]) {
            let appInfo = Gio.DesktopAppInfo.new(appID);
            this._splashes[appID] = new Window(this, appInfo);
            this._splashes[appID].connect('delete-event', Lang.bind(this, function() {
                this._emitSplashClosed(appID);
            }));
        }

        this._splashes[appID].present();
    },

    HideSplash: function(appID) {
        this._splashes[appID].rampOut();
        this._splashes[appID] = null;
    },
});

let app = new EosSpeedwagon();
app.run(null);


src/lib/speedwagon-spinner.c
/*
 * Copyright (C) 2015 Endless Mobile
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *
 * Written by:
 *     Jasper St. Pierre <jstpierre@mecheye.net>
 */

/* A hack workaround for the fact that we can't override
 * gtk_container_forall from gjs. */

#include "speedwagon-spinner.h"

struct _SpeedwagonSpinnerPrivate
{
  GSList *internal_widgets;
};
typedef struct _SpeedwagonSpinnerPrivate SpeedwagonSpinnerPrivate;

G_DEFINE_TYPE_WITH_PRIVATE (SpeedwagonSpinner, speedwagon_spinner, GTK_TYPE_BOX);

static void
speedwagon_spinner_forall (GtkContainer *container,
                           gboolean      include_internals,
                           GtkCallback   callback,
                           gpointer      callback_data)
{
  SpeedwagonSpinner *spinner = SPEEDWAGON_SPINNER (container);
  SpeedwagonSpinnerPrivate *priv = speedwagon_spinner_get_instance_private (spinner);

  GTK_CONTAINER_CLASS (speedwagon_spinner_parent_class)->forall (container, include_internals, callback, callback_data);

  if (include_internals)
    g_slist_foreach (priv->internal_widgets, (GFunc) callback, callback_data);
}

static void
speedwagon_spinner_dispose (GObject *object)
{
  SpeedwagonSpinner *spinner = SPEEDWAGON_SPINNER (object);
  SpeedwagonSpinnerPrivate *priv = speedwagon_spinner_get_instance_private (spinner);

  g_slist_free_full (priv->internal_widgets, (GDestroyNotify) g_object_unref);
}

static void
speedwagon_spinner_class_init (SpeedwagonSpinnerClass *klass)
{
  GtkContainerClass *container_class = GTK_CONTAINER_CLASS (klass);
  GObjectClass *object_class = G_OBJECT_CLASS (klass);

  container_class->forall = speedwagon_spinner_forall;
  object_class->dispose = speedwagon_spinner_dispose;
}

static void
speedwagon_spinner_init (SpeedwagonSpinner *spinner)
{
}

/**
 * speedwagon_spinner_set_internal_widgets:
 * @spinner:
 * @internal_widgets: (element-type Gtk.Widget) (transfer container):
 */
void
speedwagon_spinner_set_internal_widgets (SpeedwagonSpinner *spinner,
                                         GSList *internal_widgets)
{
  SpeedwagonSpinnerPrivate *priv = speedwagon_spinner_get_instance_private (spinner);

  priv->internal_widgets = internal_widgets;
  g_slist_foreach (priv->internal_widgets, (GFunc) g_object_ref_sink, NULL);
}


src/lib/speedwagon-spinner.h
/*
 * Copyright (C) 2015 Endless Mobile
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *
 * Written by:
 *     Jasper St. Pierre <jstpierre@mecheye.net>
 */

#pragma once

#include <gtk/gtk.h>

#define SPEEDWAGON_TYPE_SPINNER             (speedwagon_spinner_get_type ())
#define SPEEDWAGON_SPINNER(obj)             (G_TYPE_CHECK_INSTANCE_CAST ((obj), SPEEDWAGON_TYPE_SPINNER, SpeedwagonSpinner))
#define SPEEDWAGON_SPINNER_CLASS(klass)     (G_TYPE_CHECK_CLASS_CAST ((klass),  SPEEDWAGON_TYPE_SPINNER, SpeedwagonSpinnerClass))
#define SPEEDWAGON_IS_SPINNER(obj)          (G_TYPE_CHECK_INSTANCE_TYPE ((obj), SPEEDWAGON_TYPE_SPINNER))
#define SPEEDWAGON_IS_SPINNER_CLASS(klass)  (G_TYPE_CHECK_CLASS_TYPE ((klass),  SPEEDWAGON_TYPE_SPINNER))
#define SPEEDWAGON_SPINNER_GET_CLASS(obj)   (G_TYPE_INSTANCE_GET_CLASS ((obj),  SPEEDWAGON_TYPE_SPINNER, SpeedwagonSpinnerClass))

typedef struct _SpeedwagonSpinner        SpeedwagonSpinner;
typedef struct _SpeedwagonSpinnerClass   SpeedwagonSpinnerClass;

struct _SpeedwagonSpinner {
  GtkBox parent;
};

struct _SpeedwagonSpinnerClass {
  GtkBoxClass parent_class;
};

GType speedwagon_spinner_get_type (void) G_GNUC_CONST;

void speedwagon_spinner_set_internal_widgets (SpeedwagonSpinner *spinner,
                                              GSList *internal_widgets);


src/eos_speedwagon.js
// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-

const Format = imports.format;
const Lang = imports.lang;

const Config = imports.config;
const GLib = imports.gi.GLib;
const Gio = imports.gi.Gio;
const Gdk = imports.gi.Gdk;
// Needed to be imported so that get_window() returns a proper wrapper.
const GdkX11 = imports.gi.GdkX11;
const Gtk = imports.gi.Gtk;

const SpeedwagonIface = '<node> \
<interface name="com.endlessm.Speedwagon"> \
<method name="ShowSplash"> \
    <arg type="s" direction="in" name="desktopFile" /> \
</method> \
<method name="HideSplash"> \
    <arg type="s" direction="in" name="desktopFile" /> \
</method> \
<signal name="SplashClosed"> \
    <arg type="s" name="desktopFile" /> \
</signal> \
</interface> \
</node>';

const SPLASH_CIRCLE_PERIOD = 2;
const SPLASH_SCREEN_FADE_OUT = 0.2;

// This needs to match the time of the transitions in speedwagon.css
const SPLASH_SCREEN_COMPLETE_TIME = 0.2;

const SPLASH_BACKGROUND_DESKTOP_KEY = 'X-Endless-SplashBackground';
const DEFAULT_SPLASH_SCREEN_BACKGROUND = 'resource:///com/endlessm/Speedwagon/splash-background-default.jpg';

const QUIT_TIMEOUT = 15;

const Window = new Lang.Class({
    Name: 'EosSpeedwagonWindow',
    Extends: Gtk.ApplicationWindow,

    _init: function(app, appInfo) {
        this._appInfo = appInfo;

        this.parent({ application: app,
                      role: 'eos-speedwagon',
                      title: appInfo.get_name() });
        this.maximize();
        this.set_keep_above(true);

        this._buildUI();
    },

    _buildUI: function() {
        this._baseBox = new Gtk.Box({ visible: true });
        let context = this._baseBox.get_style_context();
        context.add_class('speedwagon-bg');

        this._spinner = new Gtk.Spinner({ visible: true,
                                          width_request: 220,
                                          height_request: 220,
                                          hexpand: true, vexpand: true,
                                          halign: Gtk.Align.CENTER,
                                          valign: Gtk.Align.CENTER });
        this._spinner.get_style_context().add_class('speedwagon-spinner');
        this._spinner.start();

        let spinnerBox = new Gtk.Box({ visible: true });
        spinnerBox.get_style_context().add_class('speedwagon-spinner-bg');

        let overlay = new Gtk.Overlay({ visible: true });
        overlay.add(spinnerBox);
        overlay.add_overlay(this._spinner);
        this._baseBox.add(overlay);

        let gicon = this._appInfo.get_icon();
        let image = new Gtk.Image({ visible: true,
                                    hexpand: true, vexpand: true,
                                    halign: Gtk.Align.CENTER,
                                    valign: Gtk.Align.CENTER,
                                    pixel_size: 64,
                                    gicon: gicon });
        spinnerBox.add(image);

        let bgPath;
        if (this._appInfo.has_key(SPLASH_BACKGROUND_DESKTOP_KEY)) {
            bgPath = this._appInfo.get_string(SPLASH_BACKGROUND_DESKTOP_KEY);
        } else {
            bgPath = DEFAULT_SPLASH_SCREEN_BACKGROUND;
        }

        let provider = new Gtk.CssProvider();
        provider.load_from_data(Format.vprintf('.speedwagon-bg { background-image: url("%s"); }', [bgPath]));
        context.add_provider(provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);

        this.add(this._baseBox);
    },

    _fadeOutTick: function(window, clock) {
        let time = clock.get_frame_time();
        let dt = (time - this._fadeOutStartTime) / GLib.USEC_PER_SEC;
        let opacity = 1.0 - (dt / SPLASH_SCREEN_FADE_OUT);
        if (opacity > 0)
            this.opacity = opacity;
        else
            this.destroy();
        return true;
    },
    _fadeOut: function() {
        this._fadeOutStartTime = this.get_frame_clock().get_frame_time();
        this.add_tick_callback(Lang.bind(this, this._fadeOutTick));
    },

    rampOut: function() {
        this._baseBox.get_style_context().add_class('glow');

        GLib.timeout_add(GLib.PRIORITY_DEFAULT, SPLASH_SCREEN_COMPLETE_TIME * 1000, Lang.bind(this, function() {
            this._fadeOut();
            return false;
        }));
    },
});

const EosSpeedwagon = new Lang.Class({
    Name: 'EosSpeedwagon',
    Extends: Gtk.Application,

    _init: function() {
        this.parent({ application_id: 'com.endlessm.Speedwagon',
                      flags: Gio.ApplicationFlags.IS_SERVICE,
                      inactivity_timeout: QUIT_TIMEOUT * 1000 });

        this._dbusImpl = Gio.DBusExportedObject.wrapJSObject(SpeedwagonIface, this);
        this._dbusImpl.export(Gio.DBus.session, '/com/endlessm/Speedwagon');

        this._splashes = {};
    },

    vfunc_startup: function() {
        this.parent();

        let resource = Gio.Resource.load(Config.RESOURCE_DIR + '/speedwagon.gresource');
        resource._register();

        let provider = new Gtk.CssProvider();
        provider.load_from_file(Gio.File.new_for_uri('resource:///com/endlessm/Speedwagon/speedwagon.css'));
        Gtk.StyleContext.add_provider_for_screen(Gdk.Screen.get_default(), provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
    },

    vfunc_activate: function() {
    },

    _emitSplashClosed: function(appID) {
        let connection = this._dbusImpl.get_connection();
        let info = this._dbusImpl.get_info();
        connection.emit_signal(null, this._dbusImpl.get_object_path(), info.name, 'SplashClosed', GLib.Variant.new('(s)', [appID]));
    },

    ShowSplash: function(appID) {
        if (!this._splashes[appID]) {
            let appInfo = Gio.DesktopAppInfo.new(appID);
            this._splashes[appID] = new Window(this, appInfo);
            this._splashes[appID].connect('delete-event', Lang.bind(this, function() {
                this._emitSplashClosed(appID);
                this._splashes[appID] = null;
                return false;
            }));
        }

        this._splashes[appID].present();
    },

    HideSplash: function(appID) {
        if (this._splashes[appID])
            this._splashes[appID].rampOut();

        this._splashes[appID] = null;
    },
});

function main(argv) {
    let app = new EosSpeedwagon();
    if (GLib.getenv('EOS_SPEEDWAGON_PERSIST'))
        app.hold();
    return app.run(null);
}
